storage/test/sparse_merkle/tree_test.cpp

```diff
@@ -273,13 +406,658 @@ TEST(tree_tests, multiple_updates) {
+  ASSERT_TRUE(internalNodeVersionMatches(2, batch.internal_nodes));
+  auto& [root_key, root_node] = batch.internal_nodes[0];
+  auto& [depth1_key, depth1_node] = batch.internal_nodes[1];
+  ASSERT_EQ("", root_key.path().toString());
+  ASSERT_EQ("b", depth1_key.path().toString());
+  ASSERT_TRUE(validLinkExists(root_key, root_node, depth1_key, depth1_node));
+
+  // There are no leaf updates.
+  ASSERT_EQ(0, batch.leaf_nodes.size());
+
+  // The root node still has zero leaf children and the node at depth one ("b"),
+  // has one promoted LeafChild.
+  ASSERT_EQ(0, root_node.numLeafChildren());
+  ASSERT_EQ(1, depth1_node.numLeafChildren());
+  ASSERT_TRUE(leafChildExists("key462", 1, depth1_node));
+
+  // There are 3 old internal keys (root, "b", "be") and one stale leaf
+  ASSERT_EQ(Version(2), batch.stale.stale_since_version);
+  ASSERT_EQ(3, batch.stale.internal_keys.size());
+  ASSERT_TRUE(internalKeyExists("", 1, batch.stale.internal_keys));
+  ASSERT_TRUE(internalKeyExists("b", 1, batch.stale.internal_keys));
+  ASSERT_TRUE(internalKeyExists("be", 1, batch.stale.internal_keys));
+  ASSERT_EQ(1, batch.stale.leaf_keys.size());
+  ASSERT_TRUE(leafKeyExists("key9", 1, batch.stale.leaf_keys));
+}
+
+// This is essentially a combination of the prior 3 tests in succession.
+// We omit the checks after the initial delete since they are done in the prior test.
+TEST(tree_tests, successive_removes) {
+  std::shared_ptr<TestDB> db(new TestDB);
+  Tree tree(db);
+  SetOfKeyValuePairs updates;
+  updates.emplace(Sliver("key3"), Sliver("val1"));
+  updates.emplace(Sliver("key9"), Sliver("val2"));
+  updates.emplace(Sliver("key191"), Sliver("val3"));
+  updates.emplace(Sliver("key462"), Sliver("val4"));
+  auto batch = tree.update(updates);
+  db_put(db, batch);
+
+  // Delete a LeafChild at "be". This should should trigger a promotion to "b";
+  KeysVector deletes{Sliver("key9")};
+  batch = tree.remove(deletes);
+  db_put(db, batch);
+
+  // The tree looks like this now:
+  //                Root
+  //                 |
+  //                "b"
+  //                 |
+  //                "ba"
+  //
+  // Deleting the last leaf node from "b" (key462), should not cause this tree
+  // to change, since there is a collision that forces the "ba" node to be
+  // created. In other words there is an InternalLeaf at level 0 of
+  // BatchedInternalNode "b".
+  KeysVector deletes2{Sliver("key462")};
+  batch = tree.remove(deletes2);
+  db_put(db, batch);
+
+  // "" and "b" should be updated.
+  ASSERT_EQ(2, batch.internal_nodes.size());
+  auto& [root_key, root_node] = batch.internal_nodes[0];
+  auto& [b_key, b_node] = batch.internal_nodes[1];
+  ASSERT_EQ("", root_key.path().toString());
+  ASSERT_EQ("b", b_key.path().toString());
+  ASSERT_EQ(0, root_node.numLeafChildren());
+  ASSERT_EQ(MAX_INTERNAL_PATH, root_node.numInternalChildren());
+  ASSERT_EQ(0, b_node.numLeafChildren());
+  ASSERT_EQ(MAX_INTERNAL_PATH, b_node.numInternalChildren());
```

Comment:
Could you please explain the `MAX_INTERNAL_PATH` asserts?

Reply:
It's to check that in the case of a single InternalChild at the Leaf (depth 4) of the BatchedInternalNode, pointing to another BatchedInternalNode, that there is a path from the root of the BatchedInternalNode to that leaf. It's actually a property of the BatchedInternalNode and not the tree, and as such doesn't really belong in this test. I think it's also confusing and as it doesn't add much I'm going to remove it.